Loopback HTTP redirect support for macOS

WilliamDenniss opened the pull request:
This PR adds the option to handle loopback IP HTTP redirects for macOS.  Loopback redirects are an alternative redirect choice to custom URI scheme based redirects on macOS. Both choices have their pros and cons, this gives developers the choice.

To enable this, I have introduced some third-party code (taken from an Apple sample), clearly marked as such, to provide the HTTP server functionality, modified to listen only on the loopback interface.

<!-- Reviewable:start -->
---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/openid/appauth-ios/42)

<!-- Reviewable:end -->

StevenEWright commented:
Reviewed 14 of 15 files at r1.
Review status: 14 of 15 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 64 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7wVIw4aDV3JQnwer:-KOV7wVIw4aDV3JQnwes:b-lqskf5) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L64)):*
> ```objc
> 
> @implementation AppAuthExampleViewController {
>   OIDRedirectHTTPHandler *redirectHTTPHandler;
> ```

Instance variables should start with an underscore.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 262 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7liFyy62wq2VE2ak:-KOV7liGv_wh4RGSxfsZ:b-cnq7jt) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L262)):*
> ```objc
>   // Generally you will point them at a nice page on your site that instructs them to return to the
>   // app. It's best when that page is uncluttered and to the point.
>   NSString *successURLString =
> ```

static NSString *const successURLString

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 301 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV4IQYyodIezjTXcqT:-KOV4IQYyodIezjTXcqU:bd3389g) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L301)):*
> ```objc
>                             callback:^(OIDAuthState *_Nullable authState,
>                                        NSError *_Nullable error) {
>       
> ```

Nit; Please setup Xcode to automatically trim trailing spaces ;)

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 307 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV5Ckvz-hwqdwC-0O-:-KOV5Ckvz-hwqdwC-0O0:b-8x7gcd) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L307)):*
> ```objc
>       
>       if (authState) {
>         [self setAuthState:authState];
> ```

You can move calls to setAuthState here and below out of the `if` since it's set to `authState` in both cases.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 311 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV4xLR54J9ofAaXhu3:-KOV4xLR54J9ofAaXhu4:b-lddhxy) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L311)):*
> [error localizedDescription]

localizedDescription is a property, so dot notation.

---

*[Source/macOS/OIDRedirectHTTPHandler.h, line 25 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV5XIj7h3IUjlsM8YY:-KOV5XIkRKsh_WA6CHKk:bq6uo6l) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.h#L25)):*
> ```objc
> @protocol OIDAuthorizationFlowSession;
> 
> @interface OIDRedirectHTTPHandler : NSObject
> ```

Javadoc comments on all classes, protocols, methods, properties.

---

*[Source/macOS/OIDRedirectHTTPHandler.h, line 44 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV5xr9mJOsWb4ISTwV:-KOV5xr9mJOsWb4ISTwW:b-56fd6k) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.h#L44)):*
> nullable 

NSObject init does not return `nullable instancetype` and you have no special case for returning nil.

---

*[Source/macOS/OIDRedirectHTTPHandler.h, line 49 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6B0axYpValcB-eTw:-KOV6B0bFW50gD40gI0R:b-l7ompi) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.h#L49)):*
> ```objc
>     @brief Starts listening on the loopback interface and returns a URL with the address.
>     @param error The error if an error occurred.
>     @return the URL containing the address of the server with the randomly assigned available port.
> ```

When converted to documentation, this will look like a sentence which starts with a lowercase letter "t".

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 26 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6cszws1eeh6UA-kD:-KOV6cszws1eeh6UA-kE:bau0f1w) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L26)):*
> ```objc
> @implementation OIDRedirectHTTPHandler {
>   HTTPServer *httpServ;
>   NSURL *successURL;
> ```

All instance variables should be prefixed with an underscore; here, and elsewhere.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 30 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6LjcbQQZr1sXoPQ0:-KOV6LjcbQQZr1sXoPQ1:b-995479) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L30)):*
> ```objc
> 
> - (nullable instancetype)init {
>   return [self initWithSuccessURL:nil];
> ```

Either initWithSuccessURL: should be modified to take a `nullable NSURL *` as a parameter, or this initializer should be forboden.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 33 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6hhOpR3GRHRy-GQF:-KOV6JYZc7JIdx9tfz3K:b-h06jin) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L33)):*
> ```objc
> }
> 
> - (nullable instancetype)initWithSuccessURL:(NSURL *)successURL_ {
> ```

After fixing instance variables, please change name back to `successURL`

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 36 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6mQh6Y6QWxynuv8y:-KOV6mQh6Y6QWxynuv8z:biool7u) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L36)):*
> ```objc
>   self = [super init];
>   if (self) {
>     successURL = successURL_;
> ```

NSURLs should be copied so that they don't change from underneath you; and, they generally follow the same semantics as strings.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 42 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7-6ecuMP-fPR9fyr:-KOV7-6ecuMP-fPR9fys:byaa1x5) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L42)):*
> ```objc
> 
> - (NSURL*)startHTTPListener:(NSError **)returnError {
>   // Starts a HTTP server on the loopback interface.
> ```

These comments should already be in the Javadocs.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 88 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV8U8LxtT5IlA3J-ES:-KOV8U8McwpEe2D-CTki:b-scim5c) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L88)):*
> absoluteString]

Is a property... I think there are several more of these littered around the code; like `length` a couple lines below, and elsewhere.

---

*[third_party/HTTPServer/HTTPServer.h, line 1 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7UdQgkH4T78noQBq:-KOV7UdQgkH4T78noQBr:b-nq0bb8) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/third_party/HTTPServer/HTTPServer.h#L1)):*
> ```objc
> /*
> ```

Is this the best way for us to include this third party project?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/42)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
Review status: 9 of 17 files reviewed at latest revision, 15 unresolved discussions.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 64 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7wVIw4aDV3JQnwer:-KP-aP2R2cYVVPR_Fsy6:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L64)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Instance variables should start with an underscore.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 262 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7liFyy62wq2VE2ak:-KP-aQFMy0E0lJ8Ea8EI:bdq70in) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L262)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> static NSString *const successURLString
</details>
I moved it out of the class to make it a proper constant.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 301 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV4IQYyodIezjTXcqT:-KP-amYgDSDcbc-85NLF:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L301)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Nit; Please setup Xcode to automatically trim trailing spaces ;)
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 307 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV5Ckvz-hwqdwC-0O-:-KP-anHpKG0G7z98zjax:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L307)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> You can move calls to setAuthState here and below out of the `if` since it's set to `authState` in both cases.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 311 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV4xLR54J9ofAaXhu3:-KP-anyxeU16meQS1aN1:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L311)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > [error localizedDescription]
> 
> localizedDescription is a property, so dot notation.
> 
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.h, line 25 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV5XIj7h3IUjlsM8YY:-KP-uAU_q7wlsf1rDOR3:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.h#L25)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Javadoc comments on all classes, protocols, methods, properties.
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.h, line 44 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV5xr9mJOsWb4ISTwV:-KP-apKHCHaEtVNDLXqr:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.h#L44)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > nullable 
> 
> NSObject init does not return `nullable instancetype` and you have no special case for returning nil.
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.h, line 49 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6B0axYpValcB-eTw:-KP-apwMqxu7gBLZFjsN:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.h#L49)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> When converted to documentation, this will look like a sentence which starts with a lowercase letter "t".
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 26 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6cszws1eeh6UA-kD:-KP-avvEPVctTZB6TL6u:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L26)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> All instance variables should be prefixed with an underscore; here, and elsewhere.
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 30 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6LjcbQQZr1sXoPQ0:-KP-avazUzlZOzPkDWqD:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L30)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Either initWithSuccessURL: should be modified to take a `nullable NSURL *` as a parameter, or this initializer should be forboden.
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 33 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6hhOpR3GRHRy-GQF:-KP-awcEeUFfFJ6chnoq:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L33)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> After fixing instance variables, please change name back to `successURL`
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 36 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV6mQh6Y6QWxynuv8y:-KP-ax6Zzfb4udaniXHp:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L36)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> NSURLs should be copied so that they don't change from underneath you; and, they generally follow the same semantics as strings.
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 42 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7-6ecuMP-fPR9fyr:-KP-cGspn-2CwKG8Sjpx:b-7vhfxx) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L42)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> These comments should already be in the Javadocs.
</details>
I added a javadoc comment, but I'm keeping this as it's implementation detail related to `HTTPServer` that it only uses the loopback IP, and that a random port is used when one isn't set manually.   On other platforms I would specify the loopback IP directly in this code block, so I wanted to call that out.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 88 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV8U8LxtT5IlA3J-ES:-KP-ayfBPhG-DxoGWTvK:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Source/macOS/OIDRedirectHTTPHandler.m#L88)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > absoluteString]
> 
> Is a property... I think there are several more of these littered around the code; like `length` a couple lines below, and elsewhere.
</details>
Done.

---

*[third_party/HTTPServer/HTTPServer.h, line 1 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7UdQgkH4T78noQBq:-KP-uOgKqorvmF2iInwM:bvztlsc) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/third_party/HTTPServer/HTTPServer.h#L1)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Is this the best way for us to include this third party project?
</details>
I considered some other ways, here's my analysis:
1. There is no repository or "project" per-say for this code, it's a sample from Apple's devsite. 
2. I'm not too excited about publishing and supporting a standalone HTTP server library myself.
3. The code isn't quite designed as a generic HTTP server, for example I had to change some code to use the loopback interface (I *could* make a method to support this, if I wanted to support a standalone HTTP server library).
4. The code has not been updated since 2009, so the benefits of having it as a separate dependency (like being able to sync upstream changes more easily) seem minimal.

Evaluating the pros and cons, I don't see a whole lot of benefit to spin up and support a new library just so it can be a dependency here.  Instead, we can treat it as some private implementation detail of `OIDRedirectHTTPHandler`.

The situation changes, we can always pull it out later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/42)*
<!-- Sent from Reviewable.io -->

msarda commented:
Have you thought about using https://github.com/robbiehanson/CocoaHTTPServer/ for the HTTPServer? Or do you want to avoid adding an extra dependency just for this?

WilliamDenniss commented:
I've updated the CL, pulling the Loopback HTTP server into the library proper since it's just sample code that has been modified for our needs.

I really want to avoid a dependency unless it's absolutely necessary. I reviewed CocoaHTTPServer but it looks like overkill for this simple usage.

msarda commented on Example-Mac/Source/AppAuthExampleViewController.m:
```diff
@@ -213,6 +234,89 @@ - (IBAction)authWithAutoCodeExchange:(nullable id)sender {
+      redirectHTTPHandler = nil;
+      
+      if (authState) {
+        [self setAuthState:authState];
```
For consistency, first log and then set the auth state.

msarda commented on Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h:
```diff
@@ -0,0 +1,179 @@
+    NSNetService *netService;
+}
+
+- (id)delegate;
```
These look like properties.

msarda commented on Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h:
```diff
@@ -0,0 +1,179 @@
+
+@interface HTTPServer (HTTPServerDelegateMethods)
+- (void)HTTPServer:(HTTPServer *)serv didMakeNewConnection:(HTTPConnection *)conn;
+// If the delegate implements this method, this is called  
```
Move comment above the method.

msarda commented on Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h:
```diff
@@ -0,0 +1,179 @@
+
+- (id)initWithPeerAddress:(NSData *)addr inputStream:(NSInputStream *)istr outputStream:(NSOutputStream *)ostr forServer:(HTTPServer *)serv;
+
+- (id)delegate;
```
I think this should be a weak property.

msarda commented on Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h:
```diff
@@ -0,0 +1,179 @@
+- (HTTPServerRequest *)nextRequest;
+// get the next request that needs to be responded to
+
+- (BOOL)isValid;
```
Ditto.

WilliamDenniss commented:
Review status: 4 of 10 files reviewed at latest revision, 20 unresolved discussions.

---

*[Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h, line 48 at r2](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUKYUzjQQY_M8mfR21p-r2-48:-KUO6fKjhqKzvaQ8nXRN:bi0e31w) ([raw file](https://github.com/openid/appauth-ios/blob/c667337ca1e868448870792fde30d57368266ced/Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h#L48)):*
<details><summary><i>Previously, msarda wrote…</i></summary>
> These look like properties.
</details>
Agree, but I don't want to change the interface of this class unless it's really needed since it's based on some other sample code.

---

*[Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h, line 95 at r2](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUKYUzjQQY_M8mfR21p-r2-95:-KUO86aNaT82_8wU2Qqs:b-nr3igm) ([raw file](https://github.com/openid/appauth-ios/blob/c667337ca1e868448870792fde30d57368266ced/Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h#L95)):*
<details><summary><i>Previously, msarda wrote…</i></summary>
> Move comment above the method.
</details>
Applied to all.

---

*[Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h, line 119 at r2](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUKYUzjQQY_M8mfR21p-r2-119:-KUO81qtG9fqcPuCd04E:bgbcmu3) ([raw file](https://github.com/openid/appauth-ios/blob/c667337ca1e868448870792fde30d57368266ced/Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h#L119)):*
<details><summary><i>Previously, msarda wrote…</i></summary>
> I think this should be a weak property.
</details>
This class is based on some old code which isn't using ARC.

---

*[Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h, line 129 at r2](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUKYUzjQQY_M8mfR21p-r2-129:-KUO94oSP0ogp_P_UeWu:bribihm) ([raw file](https://github.com/openid/appauth-ios/blob/c667337ca1e868448870792fde30d57368266ced/Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h#L129)):*
<details><summary><i>Previously, msarda wrote…</i></summary>
> Ditto.
</details>
As above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/42)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
Review status: 2 of 10 files reviewed at latest revision, 20 unresolved discussions.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 307 at r2](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOSCVO59U0OyESDIp1H-r2-307:-KUPLP_yy9K_dfT5UoVj:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/c667337ca1e868448870792fde30d57368266ced/Example-Mac/Source/AppAuthExampleViewController.m#L307)):*
<details><summary><i>Previously, msarda wrote…</i></summary>
> For consistency, first log and then set the auth state.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/42)*
<!-- Sent from Reviewable.io -->

StevenEWright commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 12 files at r2, 4 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 262 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7liFyy62wq2VE2ak:-KUPPNsRKozxXl-WTCae:b-18wxvd) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L262)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> I moved it out of the class to make it a proper constant.
</details>
FWIW, you don't need to change it's scope to make it a constant. It would have worked here as well, but have been more tightly scoped (to just this method.) But :thumbsup: regardless.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 307 at r2](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOSCVO59U0OyESDIp1H-r2-307:-KUPPChQKkIuSs9Cd-av:b-r0mjo2) ([raw file](https://github.com/openid/appauth-ios/blob/c667337ca1e868448870792fde30d57368266ced/Example-Mac/Source/AppAuthExampleViewController.m#L307)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> Done.
</details>
Seems back to front?

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 41 at r4](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUPQtdlQG1IZ-VY6QaV:-KUPQtdlQG1IZ-VY6QaW:b-ji5fm6) ([raw file](https://github.com/openid/appauth-ios/blob/487d7b0e58c9f1eba16ff7c7f89c47a30a6a84fe/Source/macOS/OIDRedirectHTTPHandler.m#L41)):*
> *
Space after NSURL and before *?

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 93 at r4](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUPRAhAJBfCH9zwjA-A:-KUPRAhAJBfCH9zwjA-B:bfcledh) ([raw file](https://github.com/openid/appauth-ios/blob/487d7b0e58c9f1eba16ff7c7f89c47a30a6a84fe/Source/macOS/OIDRedirectHTTPHandler.m#L93)):*
> int
I assume this is perfectly fine. You can probably just resolve this comment. But FWIW, just wanted to add a peer double-check that you've considered "int" carefully and opted not to go with long long.

---

*[Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h, line 4 at r4](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUPRW8hJRgQpqXfuH96:-KUPRW8hJRgQpqXfuH97:b-t4yeuj) ([raw file](https://github.com/openid/appauth-ios/blob/487d7b0e58c9f1eba16ff7c7f89c47a30a6a84fe/Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h#L4)):*
> ```objc
>     @brief AppAuth iOS SDK
>     @copyright
>         Copyright 2016 The AppAuth Authors.
> ```

I noticed this copyright is different than others, including the other new files.

---

*[Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.m, line 161 at r4](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUPRkG5Hy0MFRbopAB9:-KUPRkG6g6EJke2BKEzq:b-febpqw) ([raw file](https://github.com/openid/appauth-ios/blob/487d7b0e58c9f1eba16ff7c7f89c47a30a6a84fe/Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.m#L161)):*
> ```objc
>     CFHTTPMessageRef working = CFHTTPMessageCreateEmpty(kCFAllocatorDefault, TRUE);
>     CFHTTPMessageAppendBytes(working, [ibuffer bytes], [ibuffer length]);
>     
> ```

I know... I know... it's not really our code! But stray spaces!? LOL... 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/42#-:-KUPSBwrnCa7rYBU0QrM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
Review status: 5 of 11 files reviewed at latest revision, 10 unresolved discussions.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 262 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOV7liFyy62wq2VE2ak:-KUV3whZay3WpPhOUb7S:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/3d90ca972ec0719488b47de91b60635497caed80/Example-Mac/Source/AppAuthExampleViewController.m#L262)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> FWIW, you don't need to change it's scope to make it a constant. It would have worked here as well, but have been more tightly scoped (to just this method.) But :thumbsup: regardless.
</details>
Done.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 307 at r2](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KOSCVO59U0OyESDIp1H-r2-307:-KUV4D7iktmI8hLs-Ure:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/c667337ca1e868448870792fde30d57368266ced/Example-Mac/Source/AppAuthExampleViewController.m#L307)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Seems back to front?
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 41 at r4](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUPQtdlQG1IZ-VY6QaV:-KUV4PDFFtlqkOelZ2ID:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/487d7b0e58c9f1eba16ff7c7f89c47a30a6a84fe/Source/macOS/OIDRedirectHTTPHandler.m#L41)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > *
> Space after NSURL and before *?
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 93 at r4](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUPRAhAJBfCH9zwjA-A:-KUV4vTAryxXE5RVuQqP:b10f91j) ([raw file](https://github.com/openid/appauth-ios/blob/487d7b0e58c9f1eba16ff7c7f89c47a30a6a84fe/Source/macOS/OIDRedirectHTTPHandler.m#L93)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> > int
> I assume this is perfectly fine. You can probably just resolve this comment. But FWIW, just wanted to add a peer double-check that you've considered "int" carefully and opted not to go with long long.
</details>
OK, technically it's an unsigned long, even though this is a small message, I think that's more technically correct.

---

*[Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.m, line 161 at r4](https://reviewable.io:443/reviews/openid/appauth-ios/42#-KUPRkG5Hy0MFRbopAB9:-KVWFrNfn_HjC5XifEWX:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/487d7b0e58c9f1eba16ff7c7f89c47a30a6a84fe/Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.m#L161)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> I know... I know... it's not really our code! But stray spaces!? LOL... 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/42)*
<!-- Sent from Reviewable.io -->
